File: frontend/src/pages/razaoSocial.js
// Uniformiza a razão social na exibição — NÃO mexe no cadastro.
//
// O cadastro vem de origens diferentes (digitado à mão, importado da Receita,
// colado de planilha) e chega misturado: "Mark Building Gerenc. Predial Ltda."
// ao lado de "RIO BRAVO COM. DE ARMAS MUNIÇÕES E ACESS. LTDA". Empilhados como
// cabeçalho de seção, o caixa-alta grita e o outro sussurra. O padrão adotado é
// o da MKB: Caixa de Título, preposição em minúscula, abreviação com ponto.
//
// Sem JSX e sem React, como os vizinhos, para provar em Node puro.

/** Vão em minúscula quando não são a primeira palavra. */
const ATONAS = new Set(['de', 'da', 'do', 'das', 'dos', 'e', 'em', 'a', 'o', 'as', 'os', 'à', 'às']);

/**
 * Siglas e formas jurídicas que têm grafia própria.
 *
 * A lista existe porque nenhuma regra automática separa sigla de palavra:
 * "MKB" e "RIO" têm três letras cada, e só quem conhece o cliente sabe que a
 * primeira é sigla. Cliente novo cuja sigla não esteja aqui sai em Caixa de
 * Título ("Iwc" em vez de "IWC") — o conserto é acrescentar a chave abaixo.
 */
const SIGLAS = {
  LTDA: 'Ltda', ME: 'ME', EPP: 'EPP', EIRELI: 'Eireli', MEI: 'MEI',
  SA: 'S.A.', 'S/A': 'S.A.', 'S.A': 'S.A.', CIA: 'Cia', JR: 'Jr',
  MKB: 'MKB', IWC: 'IWC', BPS4: 'BPS4', CW: 'CW', SPE: 'SPE',
};

/** Sem vogal e curta: CW, MKB, BPS. Pega sigla que não está na lista fixa. */
function pareceSigla(bruto) {
  const so = bruto.replace(/[^A-Za-zÀ-ÿ0-9]/g, '');
  if (!so || so.length > 4) return false;
  if (/\d/.test(so)) return true;                       // BPS4, 3M
  return !/[AEIOUÀ-ÿaeiou]/.test(so);                   // CW, MKB, JBS
}

/** Um pedaço de palavra (entre espaço, hífen ou barra). */
function pedaco(bruto, primeiro) {
  const nu = bruto.replace(/[.,]+$/, '');               // "COM." -> "COM"
  const pontuacao = bruto.slice(nu.length);
  const chave = nu.toUpperCase();

  if (SIGLAS[chave]) return SIGLAS[chave] + pontuacao;
  // Abreviação (terminou em ponto) é palavra encurtada, nunca sigla: "Com.".
  if (!pontuacao && pareceSigla(nu)) return chave + pontuacao;
  const baixo = nu.toLocaleLowerCase('pt-BR');
  if (!primeiro && ATONAS.has(baixo)) return baixo + pontuacao;
  return baixo.charAt(0).toLocaleUpperCase('pt-BR') + baixo.slice(1) + pontuacao;
}

/**
 * Devolve a razão social pronta para exibir.
 *
 * Nome que JÁ tem minúscula é devolvido intacto (só com os espaços sobrando
 * aparados): quem cadastrou "Mark Building Gerenc. Predial Ltda." escreveu do
 * jeito que quer ver, e reprocessar só arriscaria estragar.
 */
export function formatarRazaoSocial(nome) {
  const texto = (nome || '').trim().replace(/\s+/g, ' ');
  if (!texto) return '';
  if (/[a-zà-ÿ]/.test(texto)) return texto;

  return texto.split(' ').map((palavra, i) => {
    let primeiro = i === 0;
    // A palavra inteira vem antes da repartição: "S/A" está no mapa e não pode
    // ser quebrada na barra, senão viraria "S" + "/" + "a".
    const inteira = palavra.replace(/[.,]+$/, '').toUpperCase();
    if (SIGLAS[inteira]) return SIGLAS[inteira];
    // Reparte em hífen e barra mantendo o separador: "LTDA-ME" -> "Ltda-ME".
    return palavra.split(/([-/])/).map((parte) => {
      if (parte === '-' || parte === '/') return parte;
      const saida = pedaco(parte, primeiro);
      primeiro = false;
      return saida;
    }).join('');
  }).join(' ');
}

File: backend/app/services/razao_social.py
"""Uniformiza a razão social na EXIBIÇÃO — não mexe no cadastro.

Porte do `frontend/src/pages/razaoSocial.js`. Existe em dois lugares porque a
mesma razão social aparece na tela e na mensagem que vai para o cliente, e a
mensagem é onde o nome mal escrito incomoda mais: chega na caixa de entrada
dele, com o "OLÁ, RIO BRAVO COM. DE ARMAS MUNIÇÕES E ACESS. LTDA" gritando.

Os dois arquivos precisam andar juntos; a prova compara a saída dos dois nos
mesmos casos, para uma correção num lado não passar despercebida no outro.
"""
import re

# Vão em minúscula quando não são a primeira palavra.
ATONAS = {"de", "da", "do", "das", "dos", "e", "em", "a", "o", "as", "os", "à", "às"}

# Siglas e formas jurídicas com grafia própria. Lista fixa porque nenhuma regra
# automática separa sigla de palavra: "MKB" e "RIO" têm três letras cada, e só
# quem conhece o cliente sabe que a primeira é sigla.
SIGLAS = {
    "LTDA": "Ltda", "ME": "ME", "EPP": "EPP", "EIRELI": "Eireli", "MEI": "MEI",
    "SA": "S.A.", "S/A": "S.A.", "S.A": "S.A.", "CIA": "Cia", "JR": "Jr",
    "MKB": "MKB", "IWC": "IWC", "BPS4": "BPS4", "CW": "CW", "SPE": "SPE",
}


def _parece_sigla(bruto: str) -> bool:
    """Sem vogal e curta: CW, MKB, BPS. Pega sigla fora da lista fixa."""
    so = re.sub(r"[^A-Za-zÀ-ÿ0-9]", "", bruto or "")
    if not so or len(so) > 4:
        return False
    if re.search(r"\d", so):
        return True                      # BPS4, 3M
    return not re.search(r"[AEIOUÀ-ÿaeiou]", so)


def _pedaco(bruto: str, primeiro: bool) -> str:
    nu = re.sub(r"[.,]+$", "", bruto)
    pontuacao = bruto[len(nu):]
    chave = nu.upper()
    if chave in SIGLAS:
        return SIGLAS[chave] + pontuacao
    # Abreviação (terminou em ponto) é palavra encurtada, nunca sigla: "Com.".
    if not pontuacao and _parece_sigla(nu):
        return chave + pontuacao
    baixo = nu.lower()
    if not primeiro and baixo in ATONAS:
        return baixo + pontuacao
    return (baixo[:1].upper() + baixo[1:]) + pontuacao


def formatar(nome) -> str:
    """Razão social pronta para exibir.

    Nome que JÁ tem minúscula é devolvido intacto: quem cadastrou "Mark Building
    Gerenc. Predial Ltda." escreveu do jeito que quer ver.
    """
    texto = re.sub(r"\s+", " ", (nome or "").strip())
    if not texto:
        return ""
    if re.search(r"[a-zà-ÿ]", texto):
        return texto

    saida = []
    for i, palavra in enumerate(texto.split(" ")):
        # A palavra inteira antes da repartição: "S/A" está no mapa e não pode
        # ser quebrada na barra.
        inteira = re.sub(r"[.,]+$", "", palavra).upper()
        if inteira in SIGLAS:
            saida.append(SIGLAS[inteira])
            continue
        primeiro = i == 0
        partes = []
        for parte in re.split(r"([-/])", palavra):
            if parte in ("-", "/"):
                partes.append(parte)
                continue
            partes.append(_pedaco(parte, primeiro))
            primeiro = False
        saida.append("".join(partes))
    return " ".join(saida)
